Initial implementation.

bajtos opened the pull request:
Usage (see also README):

``` js
var express = require('express');
var metrics = require('strong-express-metrics');
var app = express();

// 1. Produce the default set of metrics (Common Access Log)
app.use(metrics());

// 2. Add custom metrics
app.use(metrics(function(req, res) {
  return {
    client: {
      id: req.authInfo.app.id,
      username: req.authInfo.user.email
    },
    data: {
      // put your custom metrics here
    }
  };
});

// 3. Optionally provide a function to report the statistics.
// This step is not necessary when the app is started by 
// StrongLoop's Supervisor
metrics.onRecord(function(data) {
  // simple statsd output
  console.log('url:%s|1|c', data.request.url);
  console.log('status:%s|1|c', data.response.status);
  console.log('response-time|%s|ms', data.response.duration);
});

app.listen(3000);
```

Connect strongloop-internal/scrum-loopback#123.

This initial implementation is intentionally using a builder function. My plan is to extend the middleware factory to accept an options object (instead of builder fn) that describes how to map req/res properties to stats record and implement some sort of interpreter/compiler to convert this description into a builder function. However, that's out of scope of this PR.

/to @sam-github @rmg please review.

I am envisioning that strong-agent's probe will call `metrics.onRequest` to register an observer. Is this API good enough for that purpose?

Commit:
Initial implementation.

bajtos commented:
One more thing to discuss: should we add a pre-defined list of properties to the stats record? That way the middleware can be used OOTB without any custom function. Plus the downstream consumers (e.g. the Common Log reporter) then can safely assume that each stat record will contain at least these properties.

The downside is that if some of these properties are not required, the record is larger than necessary and thus we are wasting some space.

sam-github commented on docs.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "title": "strong-xstats",
```
do we have a naming convention for express middleware? should we? maybe `express-agent-stats` or `strong-express-stats`? Even though this can be used without agent, I don't think its terrible to put agent in the name. It is the main intended use, anybody wanting to emit records to themself could do it in a few lines without this module.

bajtos replied:
AFAIK, the only naming convention we have is to use `strong-` prefix. I am also not a big fan of long module names created as a concatenation of all relevant keywords, that's why I was trying to come up with a shorter name. Not a big deal, if you think a long descriptive name is better, then I can live with that.

The ultimate goal is to collect "API usage metrics", so perhaps the name can include some part from that?

Few candidates:
- strong-api-stats
- strong-express-agent
- strong-express-stats
- strong-express-metrics
- strong-api-metrics
- strong-api-agent

Thoughts?

sam-github replied:
Of the ones above, strong-express-metrics seems best to me. It reasonably scopes the metrics to express (there are many kinds of metrics).

bajtos replied:
Renamed.

sam-github commented on index.js:
```diff
@@ -0,0 +1,51 @@
+ *
+ * @header xstats.onRecord(observer)
+ */
+module.exports.onRecord = observe;
```
I'd call `function observe()`, `function onRecord()`, I think its easier to follow (observe sounds like the function passed in as an argument to onRecord).

But given that what you have implemented in observe() is "event emitter lite", why not just make the xstat module an event emitter? Then you get the .onRecord for free, as `.on('record'...)`, and you get removeListener, listeners, etc. Which allows you, in the future, to implement optimizations such as not bothering to emit a record when there are no listeners. And maybe to not even bother calling the record builder function if there are no listeners.

bajtos replied:
> I'd call function observe(), function onRecord(), I think its easier to follow (observe sounds like the function passed in as an argument to onRecord).

Fixed.

> But given that what you have implemented in observe() is "event emitter lite", why not just make the xstat module an event emitter?
1. Micro-optimization. Event emitter uses `fn.call(this, arg1)`, which is about 50% slower than `fn(arg1)`. And we don't need `this` object.
2. The extra API like `removeListener` and support for multiple events is not needed now.

> Which allows you, in the future, to implement optimizations such as not bothering to emit a record when there are no listeners. And maybe to not even bother calling the record builder function if there are no listeners.

That can be already implemented now via `if (!observers)`.

---

I don't care that much about the implementation. If you still think EventEmitter is a better solution, then I'll rework this patch to get it landed and move on.

sam-github commented:
per your [question](https://github.com/strongloop/xstats/pull/1#issuecomment-72059416), I think its worth taking the userRecord, and wrapping it as

```
{
  record: userRecord,
  process: {
    pid: process.pid,
    clusterId: cluster.id, // might be undefined
  },
  http {
    // all the info from: http://en.wikipedia.org/wiki/Common_Log_Format
  }
}
```

So that even without user specific info, there is always enough to create at least a basic CLF record.

Though that kind of fixed info can be collected and emitted all ready by strong-agent, if it wanted, without having any dependency on express, or any requirement for express-middleware to  be injected. It would go in https://github.com/strongloop/strongops/blob/master/agent/lib/probes/http.js#L46, where a event could be emitted (thru internal) on agent as `'http:common-log'`. 

So the main purpose of this middleware is allowing the user to report user-defined unstructured data per-request, because if there wasn't a need for user-defined, the code could all be in agent. I think that's understood, but I wanted to point it out.

And btw, I think a reasonable event name in agent for this record could be `'express:user-record'`, because it is express specific.

Most of the above is just musing. First thing is to get data to an arc viewer component, and/or a common log record.

Btw, supervisor could take a `--http-log=<file>` argument, that would subscribe to this information from agent, and write a log in this format. It could also just write the information to the the `--log` destination (where it wouldn't be common log format, but all the info would be there, scrapable by splunk or whatever log collection and analysis tools are being used). Perhaps that latter feature would be optional, triggered by `-http-log=@log` (or some magic switch).

Was logging this information from supervisor, CLF or other, a feature we are looking for? Supervisor at the moment is the component that extracts data and directs it towards thirdparties (splunk, graphite, syslog, etc), so its where such code would go if wanted.

bajtos commented:
> I think its worth taking the userRecord, and wrapping it as
> [cut]
> So that even without user specific info, there is always enough to create at least a basic CLF record.

:+1: I'll rework the PR.

> And btw, I think a reasonable event name in agent for this record could be 'express:user-record', because it is express specific.

How about `express:usage-record`?

> Was logging this information from supervisor, CLF or other, a feature we are looking for? Supervisor at the moment is the component that extracts data and directs it towards thirdparties (splunk, graphite, syslog, etc), so its where such code would go if wanted.

Possibly, depending on priorities. See https://github.com/strongloop-internal/scrum-loopback/issues/127

bajtos commented:
@sam-github I have reworked the patch, PTAL.

I decided to use a different record structure from what you have described:

``` js
{
  timestamp: Date.now(),
  client: {
    address: req.socket.address().address,
    id: undefined, // builder should override
    username: undefined // builder should override
  },
  request: {
    method: req.method,
    url: req.url
  },
  response: {
    status: res.statusCode,
    duration: res.durationInMs,
    bytes: undefined // TODO (?)
  },
  process: {
    pid: process.pid,
    workerId: cluster.worker && cluster.workerId
  },
  data: {
    // placeholder for user-provided data
  }
}
```

The user-provided builder function can override any of the pre-defined values, most importantly `client.id` and `client.username`. These two fields are expected by Common Access Log, but the values are very application-specific. Consider for example Basic auth, access-token based Auth, OAuth 2.0 or Windows/NTML auth to name few different options.

I also made the `recordBuilder` an optional argument, since the middleware produces enough useful data on its own.

bajtos commented:
I'll update the README and the PR description after we agree on the final shape of the record.

sam-github commented:
LGTM

Thoughts on reworking onRecord, so that xtats is an EventEmitter and you don't need a custom listener implementation?

bajtos commented:
> Thoughts on reworking onRecord, so that xtats is an EventEmitter and you don't need a custom listener implementation?

I responded in a line comment above, unfortunately GH UI has collapsed the thread. Cross-posting my original reply below:

> But given that what you have implemented in observe() is "event emitter lite", why not just make the xstat module an event emitter?

Micro-optimization. Event emitter uses `fn.call(this, arg1)`, which is about 50% slower than `fn(arg1)`. We don't need `this` object, so why to pay the cost?

The extra API like removeListener and support for multiple events is not needed now.

> Which allows you, in the future, to implement optimizations such as not bothering to emit a record when there are no listeners. And maybe to not even bother calling the record builder function if there are no listeners.

That can be already implemented now via `if (!observers)`.

---

I don't care that much about the implementation. If you still think EventEmitter is a better solution, then I'll rework this patch to get it landed and move on.

bajtos commented:
I have renamed the repository, updated README and cleaned up the pull request.

@sam-github I will wait with merging until we agree whether the API should be based on EventEmitter or not.

Commit:
Skip `createRecord` when there are no observers

A performance optimization. When there are no observers to receive
the metrics (usage) record, do not create the record at all.

sam-github commented:
LGTM

rmg commented:
@slnode test please
